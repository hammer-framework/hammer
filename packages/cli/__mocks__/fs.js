import { sep } from 'path'

import { isString } from 'lodash'

const fs = {
  ...jest.requireActual('fs'),
}

// This is a custom function that our tests can use during setup to specify
// what the files on the "mock" filesystem should look like when any of the
// `fs` APIs are used.
let mockFiles = {}
/** @param newMockFiles - {[filepath]: contents} */
fs.__setMockFiles = (newMockFiles) => {
  mockFiles = { ...newMockFiles }
}

fs.existsSync = (path) => {
  return isString(mockFiles[path])
}

fs.mkdirSync = () => {}

fs.readFileSync = (path) => {
  // In prisma v4.3.0, prisma format uses a Wasm module. See https://github.com/prisma/prisma/releases/tag/4.3.0.
  // We shouldn't mock this, so we'll use the real fs.readFileSync.
  if (path.includes('prisma_fmt_build_bg.wasm')) {
    return jest.requireActual('fs').readFileSync(path)
  }

  return mockFiles[path]
}

fs.readdirSync = (path) => {
  const mockedFiles = []
  Object.keys(mockFiles).forEach((mockedPath) => {
    // Is a child of the desired path
    if (mockedPath.startsWith(path)) {
      const childPath = mockedPath.substring(path.length + 1)
      // Is the child a direct child of the desired path
      if (!childPath.includes(sep)) {
        mockedFiles.push(childPath)
      }
    }
  })
  return mockedFiles
}

fs.writeFileSync = (path, contents) => {
  mockFiles[path] = contents
}

fs.unlinkSync = (path) => {
  delete mockFiles[path]
}

fs.rmSync = (path) => {
  delete mockFiles[path]
}

// @MARK this works for now but should probably be a little more full-featured
fs.rmdirSync = (path) => {
  delete mockFiles[path]
}

module.exports = fs
