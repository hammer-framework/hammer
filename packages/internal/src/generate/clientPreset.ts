import { generate } from '@graphql-codegen/cli'
import type { CodegenConfig } from '@graphql-codegen/cli'
import { addTypenameSelectionDocumentTransform } from '@graphql-codegen/client-preset'

import { getConfig, getPaths } from '@redwoodjs/project-config'

import {
  trustedDocumentsStore,
  replaceGqlTagWithTrustedDocumentGraphql,
} from './trustedDocuments'
import type { GeneratedFile } from './types'

export const shouldGenerateTrustedDocuments = (): boolean => {
  const config = getConfig()

  return config.graphql.trustedDocuments
}

export const generateClientPreset = async () => {
  let generatedFiles = []
  let clientPresetFiles = [] as string[]

  const errors: { message: string; error: unknown }[] = []

  if (!shouldGenerateTrustedDocuments()) {
    return { clientPresetFiles, trustedDocumentsStoreFile: [], errors }
  }

  const documentsGlob = `${getPaths().web.src}/**/!(*.d).{ts,tsx,js,jsx}`

  const config: CodegenConfig = {
    schema: getPaths().generated.schema,
    documents: documentsGlob,
    generates: {
      // should be graphql.d.ts
      [`${getPaths().web.base}/types/types.d.ts`]: {
        plugins: ['typescript', 'typescript-operations', 'add'],
        config: {
          enumsAsTypes: true,
          content: 'import { Prisma } from "@prisma/client"',
          placement: 'prepend',
          scalars: {
            // We need these, otherwise these scalars are mapped to any
            BigInt: 'number',
            // @Note: DateTime fields can be valid Date-strings, or the Date object in the api side. They're always strings on the web side.
            DateTime: 'string',
            Date: 'string',
            JSON: 'Prisma.JsonValue',
            JSONObject: 'Prisma.JsonObject',
            Time: 'string',
          },
          namingConvention: 'keep', // to allow camelCased query names
          omitOperationSuffix: true,
        },
      },
      [`${getPaths().web.src}/graphql/`]: {
        preset: 'client',
        presetConfig: {
          persistedDocuments: true,
        },
        documentTransforms: [addTypenameSelectionDocumentTransform],
        config: {
          // DO NOT USE documentMode: 'string',
        },
      },
    },
  }

  try {
    generatedFiles = await generate(config, true)

    clientPresetFiles = generatedFiles.map((f: GeneratedFile) => f.filename)

    const trustedDocumentsStoreFile = trustedDocumentsStore(generatedFiles)
    replaceGqlTagWithTrustedDocumentGraphql(generatedFiles)

    return {
      clientPresetFiles,
      trustedDocumentsStoreFile,
      errors,
    }
  } catch (e) {
    errors.push({
      message: 'Error: Could not generate GraphQL client preset',
      error: e,
    })

    return {
      clientPresetFiles,
      trustedDocumentsStoreFile: [],
      errors,
    }
  }
}
