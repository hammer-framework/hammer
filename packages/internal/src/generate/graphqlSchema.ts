import path from 'path'

import { generate } from '@graphql-codegen/cli'
import chalk from 'chalk'

import { getPaths } from '../paths'

export const generateGraphQLSchema = async () => {
  const rwjsPaths = getPaths()
  type GenerateResponse = { filename: string; contents: string }[]
  try {
    const f: GenerateResponse = await generate(
      {
        cwd: rwjsPaths.api.src,
        schema: [
          path.join(__dirname, '../rootGqlSchema.js'), // support loading from either compiled JS
          path.join(__dirname, '../rootGqlSchema.ts'), // or TS (for jest tests)
          'graphql/**/*.sdl.{js,ts}',
          'directives/**/*.{js,ts}',
        ],
        config: {
          scalars: {
            DateTime: 'string',
            Date: 'string',
            JSON: 'Record<string, unknown>',
            JSONObject: 'Record<string, unknown>',
            Time: 'string',
          },
        },
        generates: {
          // Note the plugin passed, we're not generating types here.
          // its subtley different to generateTypeDefGraphQL
          [rwjsPaths.generated.schema]: {
            plugins: ['schema-ast'],
          },
        },
        silent: false,
        errorsOnly: true,
      },
      true
    )
    return f[0].filename
  } catch (e: any) {
    // `generate` outputs errors which are helpful.
    // This tries to clean up the output of those errors.
    console.error(e)
    console.error(chalk.red('Error parsing SDLs or Schema'))
    for (const error of e?.errors) {
      console.error(error.details)
    }

    console.warn()
  }

  return undefined
}
