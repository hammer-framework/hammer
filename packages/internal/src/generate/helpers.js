import path from 'path'

import { getProject } from '@redwoodjs/structure'

const project = getProject()

export function generateTypeDef(filename, templatePath) {
  const { host } = project
  host.writeFileSync(
    path.join(host.paths.types, filename),
    project.host.readFileSync(templatePath)
  )
}
