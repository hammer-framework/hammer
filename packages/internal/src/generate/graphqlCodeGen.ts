import fs from 'fs'
import path from 'path'

import * as addPlugin from '@graphql-codegen/add'
import { loadCodegenConfig } from '@graphql-codegen/cli'
import { codegen } from '@graphql-codegen/core'
import type {
  Types as CodegenTypes,
  CodegenPlugin,
} from '@graphql-codegen/plugin-helpers'
import * as typescriptPlugin from '@graphql-codegen/typescript'
import * as typescriptOperations from '@graphql-codegen/typescript-operations'
import { CodeFileLoader } from '@graphql-tools/code-file-loader'
import { GraphQLFileLoader } from '@graphql-tools/graphql-file-loader'
import { loadDocuments, loadSchemaSync } from '@graphql-tools/load'
import type { LoadTypedefsOptions } from '@graphql-tools/load'
import execa from 'execa'
import {
  DocumentNode,
  InterfaceTypeDefinitionNode,
  ObjectTypeDefinitionNode,
} from 'graphql'

import { getPaths } from '../paths'
import { getTsConfigs } from '../project'

import * as rwTypescriptResolvers from './plugins/rw-typescript-resolvers'

enum CodegenSide {
  API,
  WEB,
}

export const generateTypeDefGraphQLApi = async () => {
  const filename = path.join(getPaths().api.types, 'graphql.d.ts')
  const prismaModels = getPrismaModels()
  const prismaImports = Object.keys(prismaModels).map((key) => {
    return `${key} as Prisma${key}`
  })

  const extraPlugins: CombinedPluginConfig[] = [
    {
      name: 'add',
      options: {
        content: [
          'import { Prisma } from "@prisma/client"',
          "import { MergePrismaWithSdlTypes, MakeRelationsOptional } from '@redwoodjs/api'",
          `import { ${prismaImports.join(', ')} } from '@prisma/client'`,
        ],
        placement: 'prepend',
      },
      codegenPlugin: addPlugin,
    },
    {
      name: 'print-mapped-moddels',
      options: {},
      codegenPlugin: printMappedModelsPlugin,
    },
    {
      name: 'print-prisma-models-with-implementation-chain',
      options: {},
      codegenPlugin: printPrismaTypesWithImplementationChainPlugin,
    },
    {
      name: 'typescript-resolvers',
      options: {},
      codegenPlugin: rwTypescriptResolvers,
    },
  ]

  try {
    return await runCodegenGraphQL([], extraPlugins, filename, CodegenSide.API)
  } catch (e) {
    console.error()
    console.error('Error: Could not generate GraphQL type definitions (api)')
    console.error(e)
    console.error()

    return []
  }
}

export const generateTypeDefGraphQLWeb = async () => {
  const filename = path.join(getPaths().web.types, 'graphql.d.ts')
  const options = getLoadDocumentsOptions(filename)
  const documentsGlob = './web/src/**/!(*.d).{ts,tsx,js,jsx}'

  let documents

  try {
    documents = await loadDocuments([documentsGlob], options)
  } catch {
    // No GraphQL documents present, no need to try to run codegen
    return []
  }

  const extraPlugins: CombinedPluginConfig[] = [
    {
      name: 'add',
      options: {
        content: 'import { Prisma } from "@prisma/client"',
        placement: 'prepend',
      },
      codegenPlugin: addPlugin,
    },
    {
      name: 'typescript-operations',
      options: {},
      codegenPlugin: typescriptOperations,
    },
  ]

  try {
    return await runCodegenGraphQL(
      documents,
      extraPlugins,
      filename,
      CodegenSide.WEB
    )
  } catch (e) {
    console.error()
    console.error('Error: Could not generate GraphQL type definitions (web)')
    console.error(e)
    console.error()

    return []
  }
}

/**
 * This is the function used internally by generateTypeDefGraphQLApi and generateTypeDefGraphQLWeb
 * And contains the base configuration for generating gql types with codegen
 *
 * Named a little differently to make it easier to spot
 */
async function runCodegenGraphQL(
  documents: CodegenTypes.DocumentFile[],
  extraPlugins: CombinedPluginConfig[],
  filename: string,
  side: CodegenSide
) {
  const userCodegenConfig = await loadCodegenConfig({
    configFilePath: getPaths().base,
  })

  // Merge in user codegen config with the rw built-in one
  const mergedConfig = {
    ...getPluginConfig(side),
    ...userCodegenConfig?.config?.config,
  }

  const options = getCodegenOptions(documents, mergedConfig, extraPlugins)
  const output = await codegen(options)

  fs.mkdirSync(path.dirname(filename), { recursive: true })
  fs.writeFileSync(filename, output)

  return [filename]
}

function getLoadDocumentsOptions(filename: string) {
  const loadTypedefsConfig: LoadTypedefsOptions<{ cwd: string }> = {
    cwd: getPaths().base,
    ignore: [path.join(process.cwd(), filename)],
    loaders: [new CodeFileLoader()],
    sort: true,
  }

  return loadTypedefsConfig
}

function getPrismaClient(hasGenerated = false): {
  ModelName: Record<string, string>
} {
  const localPrisma = require('@prisma/client')

  if (!localPrisma.ModelName) {
    if (hasGenerated) {
      return { ModelName: {} }
    } else {
      execa.sync('yarn rw prisma generate', { shell: true })

      // Purge Prisma Client from node's require cache, so that the newly
      // generated client gets picked up by any script that uses it
      Object.keys(require.cache).forEach((key) => {
        if (
          key.includes('/node_modules/@prisma/client/') ||
          key.includes('/node_modules/.prisma/client/')
        ) {
          delete require.cache[key]
        }
      })

      return getPrismaClient(true)
    }
  }

  return localPrisma
}

function getPrismaModels() {
  // Extract the models from the prisma client and use those to
  // set up internal redirects for the return values in resolvers.
  const localPrisma = getPrismaClient()
  const prismaModels = localPrisma.ModelName

  // This isn't really something you'd put in the GraphQL API, so
  // we can skip the model.
  if (prismaModels.RW_DataMigration) {
    delete prismaModels.RW_DataMigration
  }

  return prismaModels
}

function getPluginConfig(side: CodegenSide) {
  const prismaModels: Record<string, string> = getPrismaModels()
  Object.keys(prismaModels).forEach((key) => {
    /** creates an object like this
     * {
     *  Post: MergePrismaWithSdlTypes<PrismaPostWithImplementationChain, MakeRelationsOptional<Post, AllMappedModels>, AllMappedModels>>
     *  ...
     * }
     */
    prismaModels[
      key
    ] = `MergePrismaWithSdlTypes<Prisma${key}WithImplementationChain, MakeRelationsOptional<${key}, AllMappedModels>, AllMappedModels>`
  })

  const pluginConfig: CodegenTypes.PluginConfig &
    rwTypescriptResolvers.TypeScriptResolversPluginConfig = {
    makeResolverTypeCallable: true,
    namingConvention: 'keep', // to allow camelCased query names
    scalars: {
      // We need these, otherwise these scalars are mapped to any
      BigInt: 'number',
      // @Note: DateTime fields can be valid Date-strings, or the Date object in the api side. They're always strings on the web side.
      DateTime: side === CodegenSide.WEB ? 'string' : 'Date | string',
      Date: side === CodegenSide.WEB ? 'string' : 'Date | string',
      JSON: 'Prisma.JsonValue',
      JSONObject: 'Prisma.JsonObject',
      Time: side === CodegenSide.WEB ? 'string' : 'Date | string',
    },
    // prevent type names being PetQueryQuery, RW generators already append
    // Query/Mutation/etc
    omitOperationSuffix: true,
    showUnusedMappers: false,
    customResolverFn: getResolverFnType(),
    mappers: prismaModels,
    avoidOptionals: {
      // We do this, so that service tests can call resolvers without doing a null check
      // see https://github.com/redwoodjs/redwood/pull/6222#issuecomment-1230156868
      // Look at type or source https://shrtm.nu/2BA0 for possible config, not well documented
      resolvers: true,
    },
    contextType: `@redwoodjs/graphql-server/dist/functions/types#RedwoodGraphQLContext`,
  }

  return pluginConfig
}

export const getResolverFnType = () => {
  const tsConfig = getTsConfigs()

  if (tsConfig.api?.compilerOptions?.strict) {
    // In strict mode, bring a world of pain to the tests
    return `(
      args: TArgs,
      obj?: { root: TParent; context: TContext; info: GraphQLResolveInfo }
    ) => TResult | Promise<TResult>`
  } else {
    return `(
      args?: TArgs,
      obj?: { root: TParent; context: TContext; info: GraphQLResolveInfo }
    ) => TResult | Promise<TResult>`
  }
}

interface CombinedPluginConfig {
  name: string
  options: CodegenTypes.PluginConfig
  codegenPlugin: CodegenPlugin
}

/**
 * Codgen plugin that just lists all the SDL models and interfaces that are also mapped Prisma models
 * We use a plugin, because its possible to have Prisma models that do not have an SDL model or interface
 * so we can't just list all the Prisma models, even if they're included in the mappers object.
 *
 * Example:
 * type AllMappedModels = MaybeOrArrayOfMaybe<Post | User>
 *
 * Note that the types are SDL types, not Prisma types.
 * We do not include SDL-only types in this list.
 */
const printMappedModelsPlugin: CodegenPlugin = {
  plugin: (schema, _documents, config) => {
    // this way we can make sure relation types are not required
    const sdlTypesWhichAreMapped = Object.values(schema.getTypeMap())
      .filter((type) => {
        const kind = type.astNode?.kind
        return (
          kind === 'ObjectTypeDefinition' || kind === 'InterfaceTypeDefinition'
        )
      })
      .filter((objectDefType) => {
        const modelName = objectDefType.astNode?.name.value
        return (
          modelName && modelName in config.mappers // Only keep the mapped Prisma models
        )
      })
      .map((objectDefType) => objectDefType.astNode?.name.value)

    return `type MaybeOrArrayOfMaybe<T> = T | Maybe<T> | Maybe<T>[];\ntype AllMappedModels = MaybeOrArrayOfMaybe<${sdlTypesWhichAreMapped.join(
      ' | '
    )}>`
  },
}

/**
 * Codegen plugin that creates types for all the Prisma models with type intersections for the interfaces they implement that are also mapped Prisma models
 * We use a plugin, because we need to know the SDL types and the SDL interfaces they implement and then we need to reflect that in the constructed types made up of Prisma models.
 * This is not information we can glean from the Prisma models since there is currently no official support in Prisma for representing model inheritence (whether via Single Table Inheritence or Polymorphic associations).
 * These relationships can be represented in Prisma, but they cannot currently be represented in the Prisma schema.
 * Essentially we are constructing types out of Prisma models that reflect the SDL types and their interface/implementor relationships.
 * If there is no SDL type or interface for a Prisma model, then essentially we are creating an alias
 *
 * Examples of generated types:
 * type PrismaUserWithImplementationChain = PrismaUser
 * type PrismaCommentWithImplementationChain = PrismaComment
 * type PrismaTextCommentWithImplementationChain = PrismaTextComment & PrismaComment
 *
 * Note that the types are Prisma types, not SDL types.
 * Thus we do not include SDL-only (unmapped) types in this list.
 */
const printPrismaTypesWithImplementationChainPlugin: CodegenPlugin = {
  plugin: (schema, _documents, config) => {
    // this way we can make sure relation types are not required
    const prismaTypesWithImplementationChain = Object.values(
      schema.getTypeMap()
    )
      .filter((type) => {
        const kind = type.astNode?.kind
        return (
          kind === 'ObjectTypeDefinition' || kind === 'InterfaceTypeDefinition'
        )
      })
      .filter((objectDefType) => {
        const modelName = objectDefType.astNode?.name.value
        return (
          modelName && modelName in config.mappers // Only keep the mapped Prisma models
        )
      })
      .map((objectDefType) => {
        const modelName = objectDefType.astNode?.name.value
        const interfaces = (
          objectDefType.astNode as
            | ObjectTypeDefinitionNode
            | InterfaceTypeDefinitionNode
        ).interfaces
        if (interfaces == null || interfaces.length === 0) {
          return `type Prisma${modelName}WithImplementationChain = Prisma${modelName};`
        }

        const mappedInterfaceNames = interfaces
          .map((interfaceNode) => interfaceNode.name.value)
          .filter((interfaceName) => interfaceName in config.mappers)
          .map((interfaceName) => `Prisma${interfaceName}`)
        return `type Prisma${modelName}WithImplementationChain = Prisma${modelName} & ${mappedInterfaceNames.join(
          ' & '
        )};`
      })

    return prismaTypesWithImplementationChain.join('\n')
  },
}

function getCodegenOptions(
  documents: CodegenTypes.DocumentFile[],
  config: CodegenTypes.PluginConfig,
  extraPlugins: CombinedPluginConfig[]
) {
  const plugins = [
    { typescript: { enumsAsTypes: true } },
    ...extraPlugins.map((plugin) => ({ [plugin.name]: plugin.options })),
  ]

  const pluginMap = {
    typescript: typescriptPlugin,
    ...extraPlugins.reduce(
      (acc, cur) => ({ ...acc, [cur.name]: cur.codegenPlugin }),
      {}
    ),
  }

  const options: CodegenTypes.GenerateOptions = {
    // The typescript plugin returns a string instead of writing to a file, so
    // `filename` is not used
    filename: '',
    // `schemaAst` is used instead of `schema` if `schemaAst` is defined, and
    // `schema` isn't. In the source for GenerateOptions they have this
    // comment:
    //   Remove schemaAst and change schema to GraphQLSchema in the next major
    //   version
    // When that happens we'll have have to remove our `schema` line, and
    // rename `schemaAst` to `schema`
    schema: undefined as unknown as DocumentNode,
    schemaAst: loadSchemaSync(getPaths().generated.schema, {
      loaders: [new GraphQLFileLoader()],
      sort: true,
    }),
    documents,
    config,
    plugins,
    pluginMap,
    pluginContext: {},
  }

  return options
}
