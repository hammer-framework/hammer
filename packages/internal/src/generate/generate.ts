#!/usr/bin/env node

import { getPaths } from '@redwoodjs/project-config'

import { generateGraphQLSchema } from './graphqlSchema'
import { generatePossibleTypes } from './possibleTypes'
import { generateTypeDefs } from './typeDefinitions'

export const generate = async () => {
  const { schemaPath, errors: generateGraphQLSchemaErrors } =
    await generateGraphQLSchema()
  const { typeDefFiles, errors: generateTypeDefsErrors } =
    await generateTypeDefs()
  const { possibleTypesFiles, errors: generatePossibleTypesErrors } =
    await generatePossibleTypes()

  let files = []

  if (schemaPath !== '') {
    files.push(schemaPath)
  }

  files = [...files, ...typeDefFiles, ...possibleTypesFiles].filter(
    (x) => typeof x === 'string'
  )

  return {
    files,
    errors: [
      ...generateGraphQLSchemaErrors,
      ...generateTypeDefsErrors,
      ...generatePossibleTypesErrors,
    ],
  }
}

export const run = async () => {
  console.log('Generating...')
  console.log()

  const { files, errors } = await generate()
  const rwjsPaths = getPaths()

  for (const f of files) {
    console.log('-', f.replace(rwjsPaths.base + '/', ''))
  }
  console.log()

  if (errors.length === 0) {
    console.log('... done.')
    console.log()
    return
  }
  process.exitCode ||= 1

  console.log('... done with errors.')
  console.log()

  for (const { message, error } of errors) {
    console.error(message)
    console.log()
    console.error(error)
    console.log()
  }
}

if (require.main === module) {
  run()
}
