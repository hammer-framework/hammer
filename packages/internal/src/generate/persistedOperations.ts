import fs from 'fs'
import path from 'path'

import { GraphQLFileLoader } from '@graphql-tools/graphql-file-loader'
import { loadDocuments, loadSchemaSync } from '@graphql-tools/load'
import * as persistedOpIds from 'graphql-codegen-persisted-query-ids'

import { getPaths } from '@redwoodjs/project-config'

import { getLoadDocumentsOptions } from './graphqlCodeGen'

/**
 * Generate persisted operations from GraphQL documents.
 *
 * Persisted operations is a mechanism for preventing the execution of
 * arbitrary GraphQL operation documents.
 *
 * A 'persistedOperation' is a hash of a GraphQL operation document.
 *
 * The file 'persistedOperations.json' contains a map of persisted operation and
 * their hashes. There is a file generated for each side of the app (web and api).
 *
 * The web side is used by the client to send the persisted operation hash to the
 * server. The server then looks up the persisted operation hash in the
 * 'persistedOperations.json' file and executes the GraphQL operation document.
 *
 * The web side is stored in the 'web/src/graphql/persistedOperations.json' file.
 * The api side is stored in the 'api/lib/persistedOperations.json' file.
 *
 * By default, the persisted operations plugin follows the the
 * APQ Specification for SENDING hashes to the server.
 *
 * For example:
 *
 * ```graphql
 * query FindUserExamples {
 *    userExamples {
 *      id
 *      email
 *      name
 *    }
 *  }
 *  ```
 *
 * will become the persisted operation has for `FindUserExamples`:
 *
 * ```json
 * {
 *    "EditUserExampleById": "5649dcc8f4aaf6ecd4b43d8bd147d20b848170ce203c4af8aecaf5087c00ae7f",
 *    "UpdateUserExampleMutation": "f527c29d63ce51037ab60cd5d6b8450a9e6b79ec68197d7c6c5efe206924b35a",
 *    "CreateUserExampleMutation": "c85593858a95e9799d3350d8c51607a98f6b726b8497da5d422530bd4334f79d",
 *    "DeleteUserExampleMutation": "3ef9cc7b49ed1751649b04e6ba7c8bfce58f3e2a849886a520e5e8137b687a43",
 *    "FindUserExampleById": "a9510c2d64eb0956f90a25ca00a5364497c76adef8f2db9c93fb2e562b8a2208",
 *    "FindUserExamples": "f4e93aec76f24e336b4922baab785227ca84ee4faef3a4100d8ae9a6371baafc"
 * }
 * ```
 *
 * @see https://the-guild.dev/graphql/yoga-server/docs/features/persisted-operations
 * @see https://www.apollographql.com/docs/apollo-server/performance/apq/
 * @see https://github.com/apollographql/apollo-link-persisted-queries#apollo-engine
 **/
const generatePersistedOperationsForSide = async (
  side: 'web' | 'api'
): Promise<any> => {
  const filename =
    side === 'web'
      ? path.join(getPaths().web.graphql, 'persistedOperations.json')
      : path.join(getPaths().api.lib, 'persistedOperations.json')
  const options = getLoadDocumentsOptions(getPaths().web.src)
  const documentsGlob = './web/src/**/!(*.d).{ts,tsx,js,jsx}'

  let documents

  try {
    documents = (await loadDocuments(
      [documentsGlob],
      options
    )) as unknown as any
  } catch {
    // No GraphQL documents present, no need to try to generate persistedOperations
    return {
      persistedOperationsFiles: [],
      errors: [],
    }
  }

  const errors: { message: string; error: unknown }[] = []

  try {
    const files = []
    const pluginConfig = {
      output: side === 'web' ? 'client' : 'client',
      algorithm: 'sha256',
    } as persistedOpIds.PluginConfig
    const info = {
      outputFile: filename,
    }

    // note: the persisted ops plugin uses an older version of graphql and hence the types need some coercion
    const schema = loadSchemaSync(getPaths().generated.schema, {
      loaders: [new GraphQLFileLoader()],
      sort: true,
    }) as unknown as any

    const persistedOperations = await persistedOpIds.plugin(
      schema,
      documents,
      pluginConfig,
      info
    )

    files.push(filename)

    const output = persistedOperations.toString()

    fs.mkdirSync(path.dirname(filename), { recursive: true })
    fs.writeFileSync(filename, output)

    console.log('Persisted Operations return:', {
      persistedOperationsFiles: [filename],
      errors,
    })
    return { persistedOperationsFiles: [filename], errors } // ?
  } catch (e) {
    errors.push({
      message:
        'Error: Could not generate GraphQL persisted operations (client)',
      error: e,
    })

    return {
      persistedOperationsFiles: [],
      errors,
    }
  }
}

export const generatePersistedOperations = async () => {
  const {
    persistedOperationsFiles: persistedOperationsFilesWeb,
    errors: errorsWeb,
  } = await generatePersistedOperationsForSide('web')

  const {
    persistedOperationsFiles: persistedOperationsFilesApi,
    errors: errorsApi,
  } = await generatePersistedOperationsForSide('api')

  return {
    persistedOperationsFiles: [
      ...persistedOperationsFilesWeb,
      ...persistedOperationsFilesApi,
    ],
    errors: [...errorsWeb, ...errorsApi],
  }
}
