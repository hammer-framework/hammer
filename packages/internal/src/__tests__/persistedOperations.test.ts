import fs from 'fs'
import path from 'path'

import { getPaths } from '@redwoodjs/project-config'

import { generateGraphQLSchema } from '../generate/graphqlSchema'
import { generatePersistedOperations } from '../generate/persistedOperations'

afterEach(() => {
  delete process.env.RWJS_CWD
  jest.restoreAllMocks()
})

describe('Generate gql persisted operations from the GraphQL Schema', () => {
  test('when there are *no* union types', async () => {
    const FIXTURE_PATH = path.resolve(
      __dirname,
      '../../../../__fixtures__/example-todo-main'
    )

    process.env.RWJS_CWD = FIXTURE_PATH

    await generateGraphQLSchema()

    // note that the data for api and web side are the same
    jest
      .spyOn(fs, 'writeFileSync')
      .mockImplementation(
        (file: fs.PathOrFileDescriptor, data: string | ArrayBufferView) => {
          expect(data).toMatchSnapshot()
        }
      )

    const { persistedOperationsFiles } = await generatePersistedOperations()

    expect(persistedOperationsFiles).toHaveLength(2)

    console.debug('persistedOperationsFiles', persistedOperationsFiles)

    expect(persistedOperationsFiles[0]).toMatch(
      path.join(getPaths().web.graphql, 'persistedOperations.json')
    )
    expect(persistedOperationsFiles[1]).toMatch(
      path.join(getPaths().api.lib, 'persistedOperations.json')
    )
  })

  test('when there are union types ', async () => {
    const FIXTURE_PATH = path.resolve(
      __dirname,
      '../../../../__fixtures__/fragment-test-project'
    )

    process.env.RWJS_CWD = FIXTURE_PATH
    await generateGraphQLSchema()

    // note that the data for api and web side are the same
    jest
      .spyOn(fs, 'writeFileSync')
      .mockImplementation(
        (file: fs.PathOrFileDescriptor, data: string | ArrayBufferView) => {
          expect(data).toMatchSnapshot()
        }
      )

    const { persistedOperationsFiles } = await generatePersistedOperations()

    expect(persistedOperationsFiles).toHaveLength(2)
    expect(persistedOperationsFiles[0]).toMatch(
      path.join(getPaths().web.graphql, 'persistedOperations.json')
    )
    expect(persistedOperationsFiles[1]).toMatch(
      path.join(getPaths().api.lib, 'persistedOperations.json')
    )
  })
})
