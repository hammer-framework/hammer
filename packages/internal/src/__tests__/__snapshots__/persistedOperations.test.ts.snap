// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate gql persisted operations from the GraphQL Schema when there are *no* union types 1`] = `
"{
   "AddTodo_CreateTodo": "1863dcf516f27e2c74bc5f1f0c1ad2bd5b592a2fe8d47bf52c8ff1cce6292d0e",
   "NumTodosCell_GetCount": "a4ba88c68321773c776e350acfd874c15142ef56fc56aa96d7d69210bf2ef72c",
   "TodoListCell_GetTodos": "426d6567ac4a55dd755ae5e30a9e66013e8862069d1a32ad8d0557c630a9a133",
   "TodoListCell_CheckTodo": "e3ef938e0dea925a1ef928378fcd2fd0e81b70560ca0a4decd6bd9392d0617ac"
}"
`;

exports[`Generate gql persisted operations from the GraphQL Schema when there are *no* union types 2`] = `
"{
   "AddTodo_CreateTodo": "1863dcf516f27e2c74bc5f1f0c1ad2bd5b592a2fe8d47bf52c8ff1cce6292d0e",
   "NumTodosCell_GetCount": "a4ba88c68321773c776e350acfd874c15142ef56fc56aa96d7d69210bf2ef72c",
   "TodoListCell_GetTodos": "426d6567ac4a55dd755ae5e30a9e66013e8862069d1a32ad8d0557c630a9a133",
   "TodoListCell_CheckTodo": "e3ef938e0dea925a1ef928378fcd2fd0e81b70560ca0a4decd6bd9392d0617ac"
}"
`;

exports[`Generate gql persisted operations from the GraphQL Schema when there are *no* union types 3`] = `
"{
   "GetGroceries": "9198a2438e6e5dbcf42362657baca24494a9f6cf83a6033983d2a978c1c1c703",
   "GetProduce": "a8ee227d80bda6e1f785083aac537e8f1cd0340e0b52faaa27e18dbe4d629241"
}"
`;

exports[`Generate gql persisted operations from the GraphQL Schema when there are union types  1`] = `
"{
   "GetGroceries": "9198a2438e6e5dbcf42362657baca24494a9f6cf83a6033983d2a978c1c1c703",
   "GetProduce": "a8ee227d80bda6e1f785083aac537e8f1cd0340e0b52faaa27e18dbe4d629241"
}"
`;

exports[`Generate gql persisted operations from the GraphQL Schema when there are union types  2`] = `
"{
   "GetGroceries": "9198a2438e6e5dbcf42362657baca24494a9f6cf83a6033983d2a978c1c1c703",
   "GetProduce": "a8ee227d80bda6e1f785083aac537e8f1cd0340e0b52faaa27e18dbe4d629241"
}"
`;

exports[`Generate gql persisted operations from the GraphQL Schema when there are union types  3`] = `
"{
   "AddTodo_CreateTodo": "1863dcf516f27e2c74bc5f1f0c1ad2bd5b592a2fe8d47bf52c8ff1cce6292d0e",
   "NumTodosCell_GetCount": "a4ba88c68321773c776e350acfd874c15142ef56fc56aa96d7d69210bf2ef72c",
   "TodoListCell_GetTodos": "426d6567ac4a55dd755ae5e30a9e66013e8862069d1a32ad8d0557c630a9a133",
   "TodoListCell_CheckTodo": "e3ef938e0dea925a1ef928378fcd2fd0e81b70560ca0a4decd6bd9392d0617ac"
}"
`;

exports[`Generate gql persisted operations from the GraphQL Schema when there are union types  4`] = `
""""
Use to check whether or not a user is authenticated and is associated
with an optional set of roles.
"""
directive @requireAuth(roles: [String]) on FIELD_DEFINITION

"""Use to skip authentication checks and allow public access."""
directive @skipAuth on FIELD_DEFINITION

scalar BigInt

scalar Date

scalar DateTime

type Fruit implements Grocery {
  id: ID!

  """Seedless is only for fruits"""
  isSeedless: Boolean
  name: String!
  nutrients: String
  price: Int!
  quantity: Int!
  region: String!

  """Ripeness is only for fruits"""
  ripenessIndicators: String
  stall: Stall!
}

union Groceries = Fruit | Vegetable

interface Grocery {
  id: ID!
  name: String!
  nutrients: String
  price: Int!
  quantity: Int!
  region: String!
  stall: Stall!
}

scalar JSON

scalar JSONObject

"""About the Redwood queries."""
type Query {
  fruitById(id: ID!): Fruit
  fruits: [Fruit!]!
  groceries: [Groceries!]!

  """Fetches the Redwood root schema."""
  redwood: Redwood
  stallById(id: ID!): Stall
  stalls: [Stall!]!
  vegetableById(id: ID!): Vegetable
  vegetables: [Vegetable!]!
}

"""
The RedwoodJS Root Schema

Defines details about RedwoodJS such as the current user and version information.
"""
type Redwood {
  """The current user."""
  currentUser: JSON

  """The version of Prisma."""
  prismaVersion: String

  """The version of Redwood."""
  version: String
}

type Stall {
  fruits: [Fruit]
  id: ID!
  name: String!
  stallNumber: String!
  vegetables: [Vegetable]
}

scalar Time

type Vegetable implements Grocery {
  id: ID!

  """Pickled is only for vegetables"""
  isPickled: Boolean
  name: String!
  nutrients: String
  price: Int!
  quantity: Int!
  region: String!
  stall: Stall!

  """Veggie Family is only for vegetables"""
  vegetableFamily: String
}"
`;
